### PR TITLE
[GEP-16] Add shoot subresrouce adminkubeconfig

### PR DESCRIPTION
**Pull request by mvladev**
**How to categorize this PR?**

/area security
/kind api-change

**What this PR does / why we need it**:

The resource allows for kubeconfig requests for shoot clusters.

- New `AdminKubeconfigRequest` feature gate is added to the gardener-apiserver that enables that API.
- ~~The kubeconfig uses NIST P-256 elliptic curve for client certificate for authentication.~~
- new `--shoot-admin-kubeconfig-max-expiration` flag in gardener-apiserver allows to set maximum expiration for `AdminKubeconfigRequest`

**Which issue(s) this PR fixes**:

Part of #3823 
Related PR #3883 

**Special notes for your reviewer**:

<del>This is still a draft for early comments. I'm not sure if the option to configure max expiration as `gardener-apiserver` flag should be part of this PR or a separate one. </del>

**Release note**:

```feature operator
New `AdminKubeConfigRequest` alpha feature gate enables [AdminKubeConfigRequest](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource on shoot resources. The feature gate is disabled by default in the `gardener-apiserver` and must be explicitly enabled.
```

```feature operator
New `--shoot-admin-kubeconfig-max-expiration` flag in `gardener-apiserver` allows to specify the maximum validity duration of a credential requested to a Shoot by an `AdminKubeconfigRequest`. If an otherwise valid `AdminKubeconfigRequest` with a validity duration larger than this value is requested, a credential will be issued with a validity duration of this value. This flag is only effective when `AdminKubeConfigRequest` feature gate is enabled.
```

```feature user
If enabled in the `gardener-apiserver`, [AdminKubeConfigRequest](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) can be used to [issue a kubeconfig with cluster-admin privileges](https://github.com/gardener/gardener/blob/master/example/shoot-kubeconfig/README.md) for shoot clusters. The default expiration for such request is one hour, but the expiration time can be configured by setting `.spec.expirationSeconds` (minimum 10 minutes or 600 seconds). 
```

### CODE REVIEW

File: example/shoot-kubeconfig/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Requesting admin kubeconfigs for a Shoot cluster
+
+In order to request a kubeconfig using the [AdminKubeConfigRequest](../../docs/proosals/../proposals/16-adminkubeconfig-subresource.md) subresource on `Shoot`, the `AdminKubeConfigRequest` feature gate must be enabled on the `gardener-apiserver`:
```
Comment (rfranzke) [inline example/shoot-kubeconfig/README.md]:
```suggestion
In order to request a kubeconfig using the [AdminKubeConfigRequest](../../docs/proposals/16-adminkubeconfig-subresource.md) subresource on `Shoot`, the `AdminKubeConfigRequest` feature gate must be enabled on the `gardener-apiserver`:
```

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: Nit: Shouldn't the filename by `types_adminkubeconfigrequest.go`?

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+// AdminKubeConfigRequest is a OpenID Connect configuration that is applied
+// to a Shoot in a namespace.
+type AdminKubeConfigRequest struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Spec is the specification of this OpenIDConnect preset.
+	Spec AdminKubeConfigRequestSpec
+	// Spec is the specification of this OpenIDConnect preset.
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: Fix doc strings

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	Status AdminKubeConfigRequestStatus
+}
+
+type AdminKubeConfigRequestStatus struct {
+	KubeConfig []byte
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: doc strings

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	// ExpirationTimestamp is the time of expiration of the returned token.
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: It is no token?

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	ExpirationTimestamp metav1.Time
+}
+
+// AdminKubeConfigRequestSpec contains the Shoot selector for which
+// a specific OpenID Connect configuration is applied.
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: fix doc string

File: pkg/apis/authentication/types_adminkubeconfig_request.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+type AdminKubeConfigRequestSpec struct {
+	// ExpirationSeconds is the requested duration of validity of the request. The
+	// token issuer may return a token with a different validity duration so a
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfig_request.go]: It is no token?

File: pkg/apis/authentication/v1alpha1/types_adminkubeconfig_request.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (rfranzke) [inline pkg/apis/authentication/v1alpha1/types_adminkubeconfig_request.go]: Same comments like in `pkg/apis/authentication/types...`

File: pkg/apis/authentication/v1alpha1/types_adminkubeconfig_request.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	// Standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec is the specification of this OpenIDConnect preset.
+	Spec AdminKubeConfigRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// Spec is the specification of this OpenIDConnect preset.
+	Status AdminKubeConfigRequestStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type AdminKubeConfigRequestStatus struct {
+	KubeConfig []byte `json:"kubeConfig,omitempty" protobuf:"bytes,1,opt,name=kubeConfig"`
```
Comment (rfranzke) [inline pkg/apis/authentication/v1alpha1/types_adminkubeconfig_request.go]:
```suggestion
	Kubeconfig []byte `json:"kubeconfig,omitempty" protobuf:"bytes,1,opt,name=kubeconfig"`
```

Again, let's keep this similar to other resources that contain kubeconfigs

File: pkg/apis/authentication/validation/validation_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+)
+
+func TestValidation(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Validation Suite")
```
Comment (rfranzke) [inline pkg/apis/authentication/validation/validation_suite_test.go]:
```suggestion
	RunSpecs(t, "APIs Authentication Validation Suite")
```

File: pkg/apis/authentication/validation/validation_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+import (
+	"testing"
+	"time"
+
+	authentication "github.com/gardener/gardener/pkg/apis/authentication"
+	"github.com/gardener/gardener/pkg/apis/authentication/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
```
Comment (rfranzke) [inline pkg/apis/authentication/validation/validation_suite_test.go]:
```suggestion
	. "github.com/gardener/gardener/pkg/utils/test/matchers"

	"k8s.io/apimachinery/pkg/util/validation/field"
```

File: pkg/apis/authentication/validation/validation_suite_test.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+limitations under the License.
+*/
+
+package validation_test
+
+import (
+	"testing"
+	"time"
+
+	authentication "github.com/gardener/gardener/pkg/apis/authentication"
```
Comment (rfranzke) [inline pkg/apis/authentication/validation/validation_suite_test.go]:
```suggestion
	"github.com/gardener/gardener/pkg/apis/authentication"
```

File: pkg/features/features.go [hunk 13]
```diff
@@ -95,4 +95,9 @@ const (
 	// owner: @scheererj @docktofuture
 	// alpha: v1.22.0
 	ReversedVPN featuregate.Feature = "ReversedVPN"
+
+	// AdminKubeConfigRequest enables the AdminKubeConfigRequest endpoint on shoot resources.
+	// owner: @mvladev
+	// alpha: v1.23.0
+	AdminKubeConfigRequest featuregate.Feature = "AdminKubeConfigRequest"
```
Comment (rfranzke) [inline pkg/features/features.go]:
```suggestion
	AdminKubeconfigRequest featuregate.Feature = "AdminKubeconfigRequest"
```

This is also how k/k officials names it according to https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+	"github.com/gardener/gardener/pkg/utils/infodata"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	configlatest "k8s.io/client-go/tools/clientcmd/api/latest"
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]:
```suggestion

	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/schema"
	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
	"k8s.io/apiserver/pkg/registry/rest"
	configlatest "k8s.io/client-go/tools/clientcmd/api/latest"
```

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+	configv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+type AdminKubeConfigREST struct {
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]: doc string

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+	_ = rest.GroupVersionKindProvider(&AdminKubeConfigREST{})
+
+	gvk = schema.GroupVersionKind{
+		Group:   authenticationapiv1alpha1.SchemeGroupVersion.Group,
+		Version: authenticationapiv1alpha1.SchemeGroupVersion.Version,
+		Kind:    "AdminKubeConfigRequest",
+	}
+)
+
+func (r *AdminKubeConfigREST) New() runtime.Object {
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]: doc strings for all exported functions

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+	"context"
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	authenticationapi "github.com/gardener/gardener/pkg/apis/authentication"
+	authenticationapiv1alpha1 "github.com/gardener/gardener/pkg/apis/authentication/v1alpha1"
+	authenticationvalidation "github.com/gardener/gardener/pkg/apis/authentication/validation"
+	"github.com/gardener/gardener/pkg/apis/core"
+	corev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]:
```suggestion
	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
```

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+	corev1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]:
```suggestion
	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
```

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+	caCert, err := secrets.LoadCertificate("", caInfoData.PrivateKey, caInfoData.Certificate)
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+
+	now := r.now()
+	cc := &certConfig{
+		subject: pkix.Name{
+			CommonName:   user.GetName(),
+			Organization: []string{"system:masters"},
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]:
```suggestion
			Organization: []string{user.SystemPrivilegedGroup},
```

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+			config.CurrentContext = configName
+		}
+	}
+
+	configData, err := runtime.Encode(configlatest.Codec, config)
+	if err != nil {
+		return nil, err
+	}
+
+	out.Status.KubeConfig = []byte(configData)
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]: Why don't you use https://github.com/gardener/gardener/blob/master/pkg/utils/secrets/control_plane.go for this? It already should have everything you need?

> The kubeconfig uses NIST P-256 elliptic curve for client certificate for authentication.

Why/what for?

File: pkg/registry/core/shoot/storage/storage_suite_test.go [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestStorage(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Storage Suite")
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/storage_suite_test.go]:
```suggestion
	RunSpecs(t, "Registry Core Shoot Storage Suite")
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
>  I'm not sure if the option to configure max expiration as gardener-apiserver flag should be part of this PR or a separate one.

Hm, yes, why not/why a separate PR?

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go (hunk 20 above)
Comment (mvladev) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go]: 1. I want to switch all certificates to ECC (elliptic curve cryptography) to the lower the key size. 
2. At the moment it doesn't offer expiration time for the certificates.
3. It doesn't offer ECC
4. The kubeconfigs only have one context and cluster, in our case we want more

In general yes, we should do this, but my suggestion is to move this refactoring to a separate PR.

**Comment by petersutter** [comment]
> I'm not sure if the option to configure max expiration as gardener-apiserver flag should be part of this PR or a separate one.

Why not also have the min expiration as gardener-apiserver flag?

https://github.com/gardener/gardener/pull/3932/files#diff-28d7c8ab0bfce5b9dd923d0c423d739028593fb37d5acf4737a04cbe4102c011R33

File: docs/proposals/16-adminkubeconfig-subresource.md [hunk 22]
```diff
@@ -54,22 +54,22 @@ There are several problems with this approach:
...
 ```yaml
 apiVersion: authentication.gardener.cloud/v1alpha1
-kind: AdminKubeConfigRequest
+kind: AdminKubeconfigRequest
 spec:
   expirationSeconds: 3600
 ```
 
-Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeConfigRequest` is configured by `--shoot-admin-kubeconfig-max-expiration` flag in the `gardener-apiserver`.
+Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeconfigRequest` is configured by `--shoot-admin-kubeconfig-max-expiration` flag in the `gardener-apiserver`.
```
Comment (petersutter) [inline docs/proposals/16-adminkubeconfig-subresource.md]:
```suggestion
Where `expirationSeconds` is the validity of the certificate in seconds. In this case it would be `1 hour`. The maximum validity of a `AdminKubeconfigRequest` is configured by `--shoot-admin-kubeconfig-max-expiration` flag in the `gardener-apiserver`.
```

**Comment by mvladev** [comment]
> Why not also have the min expiration as gardener-apiserver flag?
>
> https://github.com/gardener/gardener/pull/3932/files#diff-28d7c8ab0bfce5b9dd923d0c423d739028593fb37d5acf4737a04cbe4102c011R33

I've added a new commit with the apiserver flag.

**Comment by mvladev** [comment]
Another thing - the validation of the resource won't fail if it requests longer expiration than the setting of the flag. Simply, the value would be overwritten at the rest and returned.

**Comment by petersutter** [comment]
> I've added a new commit with the apiserver flag.

I was talking about a min expiration flag :)

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go (hunk 20 above)
Comment (rfranzke) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go]: Wouldn't it be better to enhance the secret util package already now so that it supports your missing features (expiration time and multiple contexts for the kubeconfig)? You can also even start with RSA or make it configurable if required.

From my understanding, it's not a refactoring to enhance the utility library with new features if you require them for preventing code duplication. It'd indeed be a refactoring later if we merge as is.
Unless there is a good argument for proceeding as proposed, I'd vote for rather doing it now as you are already on it (we all know that once code is merged then the burden is much higher to touch it again).

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authentication/user"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	configlatest "k8s.io/client-go/tools/clientcmd/api/latest"
+	configv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+// AdminKubeconfigREST implements a RESTStorage for shoots/adminkubeconfigrequest.
```
Comment (petersutter) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go]:
```suggestion
// AdminKubeconfigREST implements a RESTStorage for shoots/adminkubeconfig.
```
isn't it `shoots/adminkubeconfig`?

**Comment by petersutter** [comment]
@mvladev doesn't it make sense to also add something like `CreateAdminKubeconfig` or similar to the shoot interface? 

```go
type ShootInterface interface {
	Create(ctx context.Context, shoot *core.Shoot, opts v1.CreateOptions) (*core.Shoot, error)
	Update(ctx context.Context, shoot *core.Shoot, opts v1.UpdateOptions) (*core.Shoot, error)
	UpdateStatus(ctx context.Context, shoot *core.Shoot, opts v1.UpdateOptions) (*core.Shoot, error)
	Delete(ctx context.Context, name string, opts v1.DeleteOptions) error
	DeleteCollection(ctx context.Context, opts v1.DeleteOptions, listOpts v1.ListOptions) error
	Get(ctx context.Context, name string, opts v1.GetOptions) (*core.Shoot, error)
	List(ctx context.Context, opts v1.ListOptions) (*core.ShootList, error)
	Watch(ctx context.Context, opts v1.ListOptions) (watch.Interface, error)
	Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts v1.PatchOptions, subresources ...string) (result *core.Shoot, err error)
	ShootExpansion
}
```

**Comment by mvladev** [comment]
> @mvladev doesn't it make sense to also add something like CreateAdminKubeconfig or similar to the shoot interface?

I think it's out of scope for now - there are no consumers of the API in the Gardener atm.

This can be done in a separate PR.

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go (hunk 20 above)
Comment (mvladev) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go]: Quite the opposite - refactoring and adding new functionality should be separate activities (PRs). If there both are added in a same PR and there is a bug in the refactoring or in the new functionality then revert would also remove the working parts.

Also the PR would grow too large to manage / review.

Comment (mvladev) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go]: After this PR is in I would start working on the refactoring. I'll even do it in stages (PRs):

- first the EC certificates
- the kubeconfig changes

Small changes are easier to review.

Comment (rfranzke) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go]: OK, then what about doing it vice-versa, i.e., first enhance/"refactor" the secret utils library and then rebase this PR on top of it? This way, it will be even easier  to review and less risk because all the changes you introduce here would not exist in the first place.

File: pkg/apis/authentication/types_adminkubeconfigrequest.go [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+type AdminKubeconfigRequestStatus struct {
+	// KubeConfig contains the kubeconfig with cluster-admin privileges for the shoot cluster.
+	Kubeconfig []byte
+	// ExpirationTimestamp is the expiration timestamp of of the returned credential.
+	ExpirationTimestamp metav1.Time
+}
+
+// AdminKubeconfigRequestSpec contains the expiration time of the kubeconfig.
+type AdminKubeconfigRequestSpec struct {
+	// ExpirationSeconds is the requested duration of validity of the request. The
```
Comment (petersutter) [inline pkg/apis/authentication/types_adminkubeconfigrequest.go]:
```suggestion
	// ExpirationSeconds is the requested validity duration of the credential. The
```

**Comment by mvladev** [comment]
I've rebased the commit on top of #4052 and fixed comments

File: pkg/apis/authentication/validation/validation.go [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+// ValidateAdminKubeconfigRequest validates a AdminKubeconfigRequest.
+func ValidateAdminKubeconfigRequest(acr *authentication.AdminKubeconfigRequest) field.ErrorList {
+	allErrs := field.ErrorList{}
+	specPath := field.NewPath("spec")
+
+	const min = 10 * time.Minute
+	if acr.Spec.ExpirationSeconds < int64(min.Seconds()) {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("expirationSeconds"), acr.Spec.ExpirationSeconds, "may not specify a duration less than 10 minutes"))
+	}
+	if acr.Spec.ExpirationSeconds > 1<<32 {
```
Comment (petersutter) [inline pkg/apis/authentication/validation/validation.go]: why don't you use a `math` constant? https://golang.org/pkg/math/#pkg-constants

Comment (mvladev) [reply pkg/apis/authentication/validation/validation.go]: Fixed in e59008f

Comment (petersutter) [reply pkg/apis/authentication/validation/validation.go]: grep for `1<<32`, there are more places you missed

File: pkg/apis/authentication/types_adminkubeconfigrequest.go [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+	// Spec is the specification of the AdminKubeconfigRequest.
+	Spec AdminKubeconfigRequestSpec
+	// Status is the status of the AdminKubeconfigRequest.
+	Status AdminKubeconfigRequestStatus
+}
+
+// AdminKubeconfigRequestStatus is the status of the AdminKubeconfigRequest containing
+// the kubeconfig and expiration of the credential.
+type AdminKubeconfigRequestStatus struct {
+	// KubeConfig contains the kubeconfig with cluster-admin privileges for the shoot cluster.
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfigrequest.go]:
```suggestion
	// Kubeconfig contains the kubeconfig with cluster-admin privileges for the shoot cluster.
```

File: pkg/apis/authentication/types_adminkubeconfigrequest.go [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+	ExpirationTimestamp metav1.Time
+}
+
+// AdminKubeconfigRequestSpec contains the expiration time of the kubeconfig.
+type AdminKubeconfigRequestSpec struct {
+	// ExpirationSeconds is the requested validity duration of the credential. The
+	// credential issuer may return a credential with a different validity duration so a
+	// client needs to check the 'expirationTimestamp' field in a response.
+	// Defaults to 1 hour.
+	// +optional
```
Comment (rfranzke) [inline pkg/apis/authentication/types_adminkubeconfigrequest.go]:
```suggestion
```

File: pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+	// Spec is the specification of the AdminKubeconfigRequest.
+	Spec AdminKubeconfigRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// Status is the status of the AdminKubeconfigRequest.
+	Status AdminKubeconfigRequestStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// AdminKubeconfigRequestStatus is the status of the AdminKubeconfigRequest containing
+// the kubeconfig and expiration of the credential.
+type AdminKubeconfigRequestStatus struct {
+	// KubeConfig contains the kubeconfig with cluster-admin privileges for the shoot cluster.
```
Comment (rfranzke) [inline pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go]:
```suggestion
	// Kubeconfig contains the kubeconfig with cluster-admin privileges for the shoot cluster.
```

File: pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+	Kubeconfig []byte `json:"kubeconfig,omitempty" protobuf:"bytes,1,opt,name=kubeconfig"`
```
Comment (rfranzke) [inline pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go]: Remove `omitempty`

File: pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// AdminKubeconfigRequestSpec contains the expiration time of the kubeconfig.
+type AdminKubeconfigRequestSpec struct {
+	// ExpirationSeconds is the requested validity duration of the credential. The
+	// credential issuer may return a credential with a different validity duration so a
+	// client needs to check the 'expirationTimestamp' field in a response.
+	// Defaults to 1 hour.
+	// +optional
+	ExpirationSeconds *int64 `json:"expirationSeconds" protobuf:"varint,1,opt,name=expirationSeconds"`
```
Comment (rfranzke) [inline pkg/apis/authentication/v1alpha1/types_adminkubeconfigrequest.go]:
```suggestion
	ExpirationSeconds *int64 `json:"expirationSeconds,omitempty" protobuf:"varint,1,opt,name=expirationSeconds"`
```

File: pkg/apis/authentication/validation/validation.go [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+func ValidateAdminKubeconfigRequest(acr *authentication.AdminKubeconfigRequest) field.ErrorList {
+	allErrs := field.ErrorList{}
+	specPath := field.NewPath("spec")
+
+	const min = 10 * time.Minute
+	if acr.Spec.ExpirationSeconds < int64(min.Seconds()) {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("expirationSeconds"), acr.Spec.ExpirationSeconds, "may not specify a duration less than 10 minutes"))
+	}
+	if acr.Spec.ExpirationSeconds > math.MaxUint32 {
+		allErrs = append(allErrs, field.TooLong(specPath.Child("expirationSeconds"), acr.Spec.ExpirationSeconds, 1<<32))
```
Comment (rfranzke) [inline pkg/apis/authentication/validation/validation.go]:
```suggestion
		allErrs = append(allErrs, field.TooLong(specPath.Child("expirationSeconds"), acr.Spec.ExpirationSeconds, math.MaxUint32))
```

File: pkg/apiserver/apiserver.go [hunk 32]
```diff
@@ -94,10 +97,23 @@ func (o *ExtraOptions) Validate() []error {
 		allErrors = append(allErrors, fmt.Errorf("--cluster-identity must be specified"))
 	}
 
+	if o.AdminKubeconfigMaxExpiration < time.Hour ||
+		o.AdminKubeconfigMaxExpiration > time.Duration(1<<32)*time.Second {
```
Comment (rfranzke) [inline pkg/apiserver/apiserver.go:101]:
```suggestion
		o.AdminKubeconfigMaxExpiration > time.Duration(math.MaxUint32)*time.Second {
```

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go [hunk 33]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	shootState := &gardencorev1alpha1.ShootState{}
+
+	err = kubernetes.GardenScheme.Convert(shootStateObj, shootState, nil)
+	if err != nil {
+		return nil, err
+	}
```
Comment (rfranzke) [inline pkg/registry/core/shoot/storage/admin_kubeconfig.go:100]: Why is this needed? Can't you work with the internal type only?

**Review by rfranzke — COMMENTED** [review]
Thanks @mvladev!
Please rebase now that #4052 is merged

File: pkg/registry/core/shoot/storage/admin_kubeconfig.go (hunk 33 above)
Comment (mvladev) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go:100]: Due to https://github.com/gardener/gardener/blob/39929464663589fb6a0b234153d4fb998c36ed87/pkg/utils/infodata/infodata.go#L61

It accepts only https://github.com/gardener/gardener/blob/95b83f3a1f3fe0c8c41dbf8193658a2d5295e4c3/pkg/apis/core/v1alpha1/helper/shootstate_list.go#L66-L67

Comment (rfranzke) [reply pkg/registry/core/shoot/storage/admin_kubeconfig.go:100]: 👍🏻 Makes sense, thank you.

**Comment by mvladev** [comment]
Rebased on top of latest commits

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by petersutter — APPROVED** [review]
/lgtm
